otherlibs/systhreads/st_stubs.c

```diff
@@ -355,15 +355,50 @@ static void caml_thread_domain_start_hook(void) {
   caml_thread_initialize_domain();
 }
 
+CAMLprim value caml_thread_join(value th);
+
+/* This hook is run when a domain shut downs. (see domains.c) */
+/* When a domain shut downs, the state must be cleared */
+/* to allow proper reuse of the domain slot the next time a domain */
+/* is started on this slot */
+/* If a program is single-domain, we mimic trunk's behavior and do not */
+/* care about ongoing thread: the program will exit. */
 static void caml_thread_domain_stop_hook(void) {
-  // This hook will clean up the initial domain's thread descriptor.
-  // The assumption is that it is the only remaining link in the threading chain
-  // (as every other threads either go through caml_thread_stop or
-  // caml_c_thread_unregister.)
-  caml_stat_free(Current_thread);
-  Current_thread = NULL;
+  caml_thread_t th, next;
+
+  /* If the program runs multiple domains, we should not let */
+  /* systhreads to hang around when a domain exit. */
+  /* If the domain is not the last one (and the last one will always */
+  /* be domain 0) we force the domain to join on every threads */
+  /* in the chaining before wrapping up, */
+  if (!caml_domain_alone()) {
+
+    th = Current_thread->next;
+
+    while (th != Current_thread) {
+      next = th->next;
+      caml_thread_join(th->descr);
+      th = next;
+    }
+
+    /* We should have joined on all threads */
+    Assert(Current_thread->next = NULL);
```

Comment:
I was worried this could block the domain for a long time *and* prevent it from participating in GC, but `caml_thread_join` calls `caml_threadstatus_wait`, which is a blocking section, so this should be all right.

Reply:
Right.
One other thing that is worth pondering over a bit, is what happens if one such _rogue_ systhread was to spawn new systhreads. (while the initial thread tries to join.)

I'm mostly convinced it is ok and the way the chaining is handled and traversed makes it that such new threads would be waited upon organically, at they would be inserted _before_ `Current_thread` (in this case, the initial thread).

I do not think we can miss such a newly spawned thread. But more eyes thinking on this specifically would be reassuring.

Reply:
> I'm mostly convinced it is ok and the way the chaining is handled and traversed makes it that such new threads would be waited upon organically, at they would be inserted before Current_thread (in this case, the initial thread).

If the thread `th` that is being joined creates new threads before terminating, they will be inserted "between" `th` and `next`, so `th = next` will skip over them.  However, when`th` terminates, they will become reachable from the thread before `th`, so maybe they will be joined eventually.

I'm more worried about the thread `th->next` terminating while we're joining `th`.  Then, `th = next` will point `th` to freed memory...

Reply:
Right, it is entirely possible that `th` ends up pointing at a freed memory space.
Maybe this loop is too simple for its own good and we would benefit from just inspecting the chaining always starting from the joining thread's entry?
Instead of looping `next`, we would always start from `Current_thread->next`, until it is `NULL`.
I think this would work fine (assuming the chaining is properly maintained by exiting threads) and would rule out possibles races.

Reply:
> we would always start from Current_thread->next, until it is NULL

You mean "until it is Current_thread" ?  Clever idea!  I think it should work fine.  The chaining of threads has to be maintained properly anyway, otherwise GC would crash when scanning their roots.

Reply:
Moreover, `Current_thread->next` cannot change while we're holding the master lock, so the test `Current_thread->next != Current_thread` is not racy.

Reply:
Indeed, `Current_thread->next == Current_thread`, not `NULL`, my bad!
I just pushed the relevant change, as well as removing the assertion under this specific loop, that was not only invalid (`= NULL` rather than `== Current_thread`), but since this is also the precondition for exiting the loop now, it is not useful.